crates/wasm-compose/src/composer.rs

```diff
@@ -56,6 +56,17 @@ impl<'a> CompositionGraphBuilder<'a> {
         let mut graph = CompositionGraph::new();
         graph.add_component(Component::from_file(ROOT_COMPONENT_NAME, root_path)?)?;
 
+        // If definition components are provided, parse them, populating
+        // them into dependencies
+        for def in &config.definitions {
+            let component = Component::from_file("", root_path.parent().unwrap().join(def))?;
+            for (expt_name, _) in &component.exports {
+                let mut dep_component = component.clone();
+                dep_component.name = expt_name.to_string();
+                graph.add_component(dep_component)?;
+            }
+        }
```

Comment:
I don't think this is the right way to do this as you're duplicating the component in the graph unnecessarily.

Additionally, this would result in multiple defined components in the composition if the root imports more than one instance from the same definition component. I confirmed this by having the test `root.wat` import more than one interface exported by `defs.wat` and it results in having `defs.wat` defined twice in the composed component.

Let me noodle on the best way to do this.

Reply:
I suppose the thing that works here that surprised me was that we could define:

```
dependencies:
  'wasi:cli-base/environment': wasi_virt.wasm
```

and have that correctly wire up `wasi:cli-base/environment` both as the dependency name AND the export name. Which seems to effectively be a sort of flattening.

Strictly speaking I would have expected something like:

```
dependencies:
  'wasi:cli-base/environment': 'wasi_virt.wasm#wasi:cli-base/environment'
```

with `#` meaning "export name", as the flattening that is actually happening here.